packages/preferences/src/browser/folders-preferences-provider.ts

```diff
@@ -21,87 +21,113 @@ import { FolderPreferenceProvider, FolderPreferenceProviderFactory } from './fol
-                    if (!this.existsProvider(root.uri)) {
-                        this.providers.push(provider);
-                    } else {
-                        provider.dispose();
+                this.resourceKinds.forEach(kind => {
+                    if (!this.existsProvider(root.uri, kind)) {
+                        const provider = this.createFolderPreferenceProvider(root, kind);
+                        this.setProvider(provider, kind);
                     }
-                }
+                });
             }
 
-            const numProviders = this.providers.length;
-            for (let ind = numProviders - 1; ind >= 0; ind--) {
-                const provider = this.providers[ind];
-                if (roots.findIndex(r => !!provider.uri && r.uri === provider.uri.toString()) < 0) {
-                    this.providers.splice(ind, 1);
-                    provider.dispose();
+            this.resourceKinds.forEach(kind => {
+                const providers = this.providersByKind.get(kind);
+                if (!providers || providers.length === 0) {
+                    return;
                 }
-            }
+                const numProviders = providers.length;
+                for (let i = numProviders - 1; i >= 0; i--) {
+                    const provider = providers[i];
+                    if (!this.existsRoot(roots, provider)) {
+                        providers.splice(i, 1);
+                        provider.dispose();
+                    }
+                }
+            });
         });
     }
 
-    private existsProvider(folderUri: string): boolean {
-        return this.providers.findIndex(p => !!p.uri && p.uri.toString() === folderUri) >= 0;
+    private existsProvider(folderUri: string, kind: ResourceKind): boolean {
+        const providers = this.providersByKind.get(kind);
+        return !!providers && providers.some(p => !!p.uri && p.uri.toString() === folderUri);
+    }
+
+    private existsRoot(roots: FileStat[], provider: FolderPreferenceProvider): boolean {
+        return roots.some(r => !!provider.uri && r.uri === provider.uri.toString());
     }
 
     // tslint:disable-next-line:no-any
     getPreferences(resourceUri?: string): { [p: string]: any } {
-        const numProviders = this.providers.length;
-        if (resourceUri && numProviders > 0) {
-            const provider = this.getProvider(resourceUri);
-            if (provider) {
-                return provider.getPreferences();
-            }
+        if (!resourceUri) {
+            return {};
         }
-        return {};
+
+        const prefProvider = this.getProvider(resourceUri, 'settings');
+        const prefs = prefProvider ? prefProvider.getPreferences() : {};
+
+        const launchProvider = this.getProvider(resourceUri, 'launch');
+        const launch = launchProvider ? launchProvider.getPreferences() : {};
+
+        const result = Object.assign({}, prefs, launch);
```

Comment:
Should not the last arg be `{ launch }` and only if launch is defined?

Reply:
No it shouldn't, see https://github.com/theia-ide/theia/pull/4743/files#diff-ef3f813f6d83ce1df009cf90a504ac82R33